### ISSUE TITLE
'transfer failed' state has no way to get out of it

### ISSUE THREAD

**Issue by indirectlylit** [learningequality/kolibri#2207]
The 'transfer failed' state has no mechanism for proceeding - the user is stuck here:

![image](https://user-images.githubusercontent.com/2367265/30554782-87822aa0-9c5a-11e7-9282-85755ba79c46.png)

cc @jonboiser

**Comment by rtibbles** [comment]
Hrm, guess my failure state emitting has failed somewhere?

**Comment by indirectlylit** [comment]
might just be a front-end problem - not sure

**Comment by jonboiser** [comment]
Yeah, I'm not sure how exactly this watch will work

```
      taskHasFailed: (/* no val here */) => {
        if (this.taskHasFailed) {
          this.$emit('taskfailed');
        }
      },

```

`this.taskHasFailed` might lag behind

**Comment by indirectlylit** [comment]
oh - I totally missed all this code. Some weird stuff going on.

The parent is passing in `firstTask` as a prop which includes a `status` attribute which drives computed prop `taskHasFailed` which is watched and emits `taskfailed` event which is listened to by the parent!

Since the parent has all info available upfront in `firstTask` the event- and watch-driven logic should be unnecessary.

**Comment by indirectlylit** [comment]
ah - good catch.

(In general though both `watch` and `$emit` should be avoided when possible since they lead to non-declarative, event-driven views. I'm not too familiar with this code so can't really comment on what an alternative would look like here)

**Comment by DXCanas** [comment]
@indirectlylit It doesn't look like it'd be too hard to make this state-driven. I think the reason it was written with `$emit`s was to have the `task-progress` component be "pure". Events are being used to notify the parent when the leg work needs to be done.

An alternative (which I think you'd favor?) would be to have the information carried by the computed props (like `taskHasFailed`) as state properties, then have `task-progress` use a getter (or prop, i guess) to read those and know when to react. The parent would also have a getter for that information, and use that to know when to run its notification-related logic... this way the parent doesn't need updates from the parent, though the connection between the child and the data becomes a little more ambiguous. 

I don't much mind its current implementation, except for the way `v-bind` is being used to attach the properties of `firstTask` to the child component without scoping them to a single prop. :man_shrugging: . Made it a kinda confusing, as `status` and `percentage` were never explicitly attached to the child. 

I think we should bring in the quickest fix for now, though.

**Comment by indirectlylit** [comment]
> I think we should bring in the quickest fix for now, though.

agreed

>  I think the reason it was written with $emits was to have the task-progress component be "pure". Events are being used to notify the parent when the leg work needs to be done.

I don't quite understand this. My (possibly naive) perspective is that there shouldn't be any 'triggering' or even 'work to be done' - those are event-driven, imperative constructs.

In particular, `showTaskCompleteNotification()` and `showTaskFailedNotification()` are implemented as imperative methods, and `notification` is implemented as local state. There might be a reason for this but it's not clear to me at first glance.

Instead, the two methods could be removed and `notification` could be a computed property which reactively returns one of the notification types based on the current value of `this.firstTask`

**Comment by DXCanas** [comment]
@jonboiser why reopoen?

**Comment by jonboiser** [comment]
@DXCanas I fixed a different problem than OP in #2230. The issue described in the OP is specific to errors that occur when deleting channels, which is still in play.

**Comment by indirectlylit** [comment]
Reproduced by @jtamiace during the hack session: 

After canceling import, trying to delete the leftover ghost channel leads to two error messages: "transfer failed. please try again" and "there was a problem deleting this channel" And... I can't bring in channels anymore and have to restart the entire installation

* Start import for a large channel
* cancel process at early stage
* try to delete the "ghost" channel

At one point we'd thought maybe this came from hitting the 'delete' button multiple times, and a mitigating strategy was to ensure all UI is disabled during the course of an operation.

However we haven't proven that theory as far as I know.

cc @aronasorman
